chanfitness/chanevent.go

```diff
@@ -29,74 +28,152 @@ func (e eventType) String() string {
-type chanEventLog struct {
-	// channelPoint is the outpoint for the channel's funding transaction.
-	channelPoint wire.OutPoint
+// peerLog tracks events for a peer and its channels. If we currently have no
+// channels with the peer, it will simply track its current online state. If we
+// do have channels open with the peer, it will track the peer's online and
+// offline events so that we can calculate uptime for our channels. A single
+// event log is used for these online and offline events, and uptime for a
+// channel is calculated by examining a subsection of this log.
+type peerLog struct {
+	// online stores whether the peer is currently online.
+	online bool
 
-	// peer is the compressed public key of the peer being monitored.
-	peer route.Vertex
-
-	// events is a log of timestamped events observed for the channel.
-	events []*channelEvent
+	// events is a log of timestamped events observed for the peer.
+	onlineEvents []*event
 
 	// clock allows creation of deterministic unit tests.
 	clock clock.Clock
 
+	// channels contains a set of currently open channels. Channels will be
+	// added and removed from this map ad they are opened and closed.
+	channels map[wire.OutPoint]*channelInfo
+}
+
+// newPeerLog creates a log for a peer.
+func newPeerLog(clock clock.Clock) *peerLog {
+	return &peerLog{
+		clock:    clock,
+		channels: make(map[wire.OutPoint]*channelInfo),
+	}
+}
+
+// channelInfo contains information about a channel.
+type channelInfo struct {
 	// openedAt tracks the first time this channel was seen. This is not
 	// necessarily the time that it confirmed on chain because channel
 	// events are not persisted at present.
 	openedAt time.Time
+}
 
-	// closedAt is the time that the channel was closed. If the channel has
-	// not been closed yet, it is zero.
-	closedAt time.Time
+func newChannelInfo(openedAt time.Time) *channelInfo {
+	return &channelInfo{
+		openedAt: openedAt,
+	}
 }
 
-// newEventLog creates an event log for a channel with the openedAt time set.
-func newEventLog(channelPoint wire.OutPoint, peer route.Vertex,
-	clock clock.Clock) *chanEventLog {
+// event records a peer online or offline event in the log.
+func (p *peerLog) onlineEvent(online bool) {
+	p.online = online
 
-	eventlog := &chanEventLog{
-		channelPoint: channelPoint,
-		peer:         peer,
-		clock:        clock,
-		openedAt:     clock.Now(),
+	// If we have no channels currently open with the peer, we do not want
+	// to commit resources to tracking their online state beyond a simple
+	// online boolean, so we exit early.
+	if p.channelCount() == 0 {
+		return
 	}
 
-	return eventlog
+	p.addOnlineEvent(online, p.clock.Now())
 }
 
-// close sets the closing time for an event log.
-func (e *chanEventLog) close() {
-	e.closedAt = e.clock.Now()
+// addOnlineEvent records an online event in our event log.
+func (p *peerLog) addOnlineEvent(online bool, time time.Time) {
```

Comment:
nit: a bit confusing that `addOnlineEvent` records both on and offline events, maybe consider changing the godoc to reflect this.

Reply:
updated to `addEvent` and updated doc, since we also have an `onlineEvent` function, the two can be confusing.